src/renderer/screens/Preferences/MyKeyboard.js

```diff
@@ -31,52 +31,51 @@ import KeyboardLEDPreferences from "./keyboard/KeyboardLEDPreferences";
 import PluginPreferences from "./keyboard/PluginPreferences";
 
 const MyKeyboardPreferences = (props) => {
-  const oneShotCancelKeyCode = 53630;
-  const escKeyCode = 41;
-
   const [modified, setModified] = useState(false);
-  const [defaultLayer, setDefaultLayer] = useState(126);
-  const [ledBrightness, setLedBrightness] = useState(255);
-  const [ledIdleTimeLimit, setLedIdleTimeLimit] = useState(0);
-  const [escOneShot, setEscOneShot] = useState(true);
+  const [saveCallbacks, setSaveCallbacks] = useState({});
 
   const { t } = useTranslation();
   const globalContext = useContext(GlobalContext);
   const [activeDevice] = globalContext.state.activeDevice;
 
+  const registerModifications = (command, args) => {
```

Comment:
register what kind of modifications?  is this onPreferenceChanged?

Reply:
It's... something like that. This method gets passed down to MyKeyboard's children, so they can hook into the [saveKeymapChanges](https://github.com/keyboardio/Chrysalis/blob/35bec2dd16c5392af9bdf4d0ca3710c11c2659fc/src/renderer/screens/Preferences/MyKeyboard.js#L50-L59) method (which I need to rename, whoops).

Said function, which really should be `saveChanges` goes and calls the registered commands. This is how we can send up what needs to be saved. We need to send those up, because the idea is to have a single "Save" button on the screen, rather than many, in each section. But we also want to have the preference-specific state as close to the component that actually renders it as possible. So we let the leaf handle all the state, data fetching, etc, because MyKeyboard doesn't need to know anything about that. It does have to know how to save stuff, so we pass those up.

Not convinced this is the best way to do that. Perhaps there's some more react-y way to do that. Channels, maybe? I suppose I could use channels.

Reply:
I tried callbacks, first just having a function in the leaf that did a focus call - that didn't work, because by the time we call the save, our state will be clobbered. We can close around that, but then we're pretty much where we are now, just with an additional layer inbetween.

Like, instead of `onSaveChanges("led_mode.default", mode)`, we'd do `onSaveChanges((focus) => focus.command("led_mode.default", mode))`. Or if we want to hide focus, `onSaveChanges(() => {"led_mode.default": mode});`. I don't think either of those would be an improvement. `onSaveChanges("led_mode.default", mode)` feels more declarative than either of the others...